diff --git a/src/lib/ecore_wayland/ecore_wl_input.c b/src/lib/ecore_wayland/ecore_wl_input.c
index 9bab9a1..9586e00 100644
--- a/src/lib/ecore_wayland/ecore_wl_input.c
+++ b/src/lib/ecore_wayland/ecore_wl_input.c
@@ -220,7 +220,7 @@ _ecore_wl_input_cursor_update(void *data)
    cursor_image = input->cursor->images[input->cursor_current_index];
    if (!cursor_image) return ECORE_CALLBACK_RENEW;
 
-   if ((buffer = wl_cursor_image_get_buffer(cursor_image)))
+   if ((buffer = wl_cursor_image_get_buffer(cursor_image)) && input->cursor_surface)
      {
         ecore_wl_input_pointer_set(input, input->cursor_surface,
                                    cursor_image->hotspot_x,
