diff --git a/src/client/qwaylandinputdevice.cpp b/src/client/qwaylandinputdevice.cpp
index 39d1705..3e715c3 100644
--- a/src/client/qwaylandinputdevice.cpp
+++ b/src/client/qwaylandinputdevice.cpp
@@ -874,6 +874,7 @@ void QWaylandInputDevice::Touch::touch_down(uint32_t serial,
     mParent->mSerial = serial;
     mFocus = QWaylandWindow::fromWlSurface(surface);
     mParent->handleTouchPoint(id, wl_fixed_to_double(x), wl_fixed_to_double(y), Qt::TouchPointPressed);
+    touch_frame();
 }
 
 void QWaylandInputDevice::Touch::touch_up(uint32_t serial, uint32_t time, int32_t id)
@@ -895,6 +896,7 @@ void QWaylandInputDevice::Touch::touch_motion(uint32_t time, int32_t id, wl_fixe
 {
     Q_UNUSED(time);
     mParent->handleTouchPoint(id, wl_fixed_to_double(x), wl_fixed_to_double(y), Qt::TouchPointMoved);
+    touch_frame();
 }
 
 void QWaylandInputDevice::Touch::touch_cancel()
