From 547125214bcb20beb0b60f3709122a94c4343e71 Mon Sep 17 00:00:00 2001
From: Gerome <gburlats@gdebian>
Date: Wed, 16 Dec 2015 11:27:22 +0100
Subject: [PATCH] Fix error when using fbdev backend with gal2d

---
 src/compositor-fbdev.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/compositor-fbdev.c b/src/compositor-fbdev.c
index 9d18c45..afb87ef 100644
--- a/src/compositor-fbdev.c
+++ b/src/compositor-fbdev.c
@@ -593,8 +593,9 @@ fbdev_output_create(struct fbdev_compositor *compositor,
 			fprintf(stderr, "failed to create window\n");
 			 return 0;
 		}
-		setenv("FB_FRAMEBUFFER_0", fbenv, 1);
-
+		if (fbenv != NULL) {
+			setenv("FB_FRAMEBUFFER_0", fbenv, 1);
+		}
 		if (gal2d_renderer->output_create(&output->base,
 					output->display,
 					(NativeWindowType)output->window) < 0) {
-- 
2.1.4

