From 120e4a232cd5af5366559073c50b2d617d54ec2c Mon Sep 17 00:00:00 2001
From: Nicolas Aguirre <aguirre.nicolas@gmail.com>
Date: Thu, 24 Apr 2014 08:38:58 +0000
Subject: [PATCH] Fix build with imx6 vivante egl drivers

---
 .../EGLWLMockNavigation/include/OpenGLES2App.h                         | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/ivi-layermanagement-examples/EGLWLMockNavigation/include/OpenGLES2App.h b/ivi-layermanagement-examples/EGLWLMockNavigation/include/OpenGLES2App.h
index fbc69a6..4c24f14 100644
--- a/ivi-layermanagement-examples/EGLWLMockNavigation/include/OpenGLES2App.h
+++ b/ivi-layermanagement-examples/EGLWLMockNavigation/include/OpenGLES2App.h
@@ -20,6 +20,9 @@
 #ifndef _OPENGLES2APP_H
 #define _OPENGLES2APP_H
 
+#define EGL_API_FB 1
+#define LINUX 1
+
 #include "wayland-client.h"
 #include "wayland-egl.h"
 #include <GLES2/gl2.h>
-- 
1.8.1.2

